apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/HighPerformancePubsub.cs

```diff
@@ -0,0 +1,163 @@
+            long recvSum = 0L; // Sum of bytes of received messages
+            Task subTask = bulkSubscriber.StartAsync((msg, ct) =>
+            {
+                var localRecvCount = Interlocked.Increment(ref recvCount);
+                if (localRecvCount <= initialNackCount)
+                {
+                    return Task.FromResult(HiPerfSubscriber.Reply.Nack);
+                }
+                Interlocked.Add(ref recvSum, msg.Data.Sum(x => (long)x));
+                if (localRecvCount >= messageCount + initialNackCount)
+                {
+                    // Test finished, so stop subscriber
+                    Task unused = bulkSubscriber.StopAsync(TimeSpan.FromSeconds(15));
+                }
+                // ACK all messages
+                return Task.FromResult(HiPerfSubscriber.Reply.Ack);
+            });
+
+            // Publish
+            var rnd = new Random(1234);
+            var activePubs = new HashSet<Task>();
+            int sentCount = 0;
+            long sentSum = 0L; // Sum of bytes of sent messages
+
+            // Watchdog to report progress and fail test on deadlock
+            int prevSentCount = -1;
+            int prevRecvCount = -1;
+            CancellationTokenSource watchdogCts = new CancellationTokenSource();
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    await Task.Delay(TimeSpan.FromSeconds(1), watchdogCts.Token).ConfigureAwait(false);
+                    var localSentCount = Interlocked.Add(ref sentCount, 0);
+                    var localRecvCount = Interlocked.Add(ref recvCount, 0);
+                    if (prevSentCount == localSentCount && prevRecvCount == localRecvCount)
+                    {
+                        // Deadlock, shutdown subscriber, and cancel
+                        Console.WriteLine("Deadlock detected. Cancelling test");
+                        bulkSubscriber.StopAsync(new CancellationToken(true));
+                        watchdogCts.Cancel();
+                    }
+                    prevSentCount = localSentCount;
+                    prevRecvCount = localRecvCount;
+                    Console.WriteLine($"Sent: {localSentCount}; Recv: {localRecvCount}");
+                }
+            });
+
+            for (int i = 0; i < messageCount; i++)
+            {
+                if (watchdogCts.IsCancellationRequested)
+                {
+                    Assert.True(false, "Test cancelled by watchdog");
+                }
+                var msgSize = rnd.Next(minMessagesize, maxMessageSize + 1);
+                var msg = new byte[msgSize];
+                rnd.NextBytes(msg);
+                sentSum += msg.Sum(x => (long)x);
+                // Send message, and record Task
+                var pubTask = bulkPublisher.PublishAsync(msg);
+                Interlocked.Increment(ref sentCount);
+                lock (activePubs) activePubs.Add(pubTask);
```

Comment:
Done.
There isn't a concurrent HashMap. There's a ConcurrentBag, but it's not the same thing.